File: split/showactivities.pas
procedure showactivities;
var
  p: 0..nestsize;
  m: -203..203;
  a: memoryword;
  q, r: halfword;
  t: integer;
begin
  nest[nestptr] := curlist;
  printnl(338);
  println;
  for p := nestptr downto 0 do
  begin
    m := nest[p].modefield;
    a := nest[p].auxfield;
    printnl(363);
    printmode(m);
    print(364);
    printint(abs(nest[p].mlfield));
    if m = 102 then
      if nest[p].pgfield <> 8585216 then
      begin
        print(365);
        printint(nest[p].pgfield mod 65536);
        print(366);
        printint(nest[p].pgfield div 4194304);
        printchar(44);
        printint((nest[p].pgfield div 65536) mod 64);
        printchar(41);
      end;
    if nest[p].mlfield < 0 then
      print(367);
    if p = 0 then
    begin
      {986:}
if 29998 <> pagetail then
      begin
        printnl(979);
        if outputactive then
          print(980);
        showbox(mem[29998].hh.rh);
        if pagecontents > 0 then
        begin
          printnl(981);
          printtotals;
          printnl(982);
          printscaled(pagesofar[0]);
          r := mem[30000].hh.rh;
          while r <> 30000 do
          begin
            println;
            printesc(330);
            t := mem[r].hh.b1 - 0;
            printint(t);
            print(983);
            if eqtb[5318 + t].int = 1000 then
              t := mem[r + 3].int
            else
              t := xovern(mem[r + 3].int, 1000) * eqtb[5318 + t].int;
            printscaled(t);
            if mem[r].hh.b0 = 1 then
            begin
              q := 29998;
              t := 0;
              repeat
                q := mem[q].hh.rh;
                if (mem[q].hh.b0 = 3) and (mem[q].hh.b1 = mem[r].hh.b1) then
                  t := t + 1;
              until q = mem[r + 1].hh.lh;
              print(984);
              printint(t);
              print(985);
            end;
            r := mem[r].hh.rh;
          end;
        end;
      end{:986}
;
      if mem[29999].hh.rh <> 0 then
        printnl(368);
    end;
    showbox(mem[nest[p].headfield].hh.rh);
    {219:}
case abs(m) div (101) of
      0:
      begin
        printnl(369);
        if a.int <= -65536000 then
          print(370)
        else
          printscaled(a.int);
        if nest[p].pgfield <> 0 then
        begin
          print(371);
          printint(nest[p].pgfield);
          print(372);
          if nest[p].pgfield <> 1 then
            printchar(115);
        end;
      end;
      1:
      begin
        printnl(373);
        printint(a.hh.lh);
        if m > 0 then
          if a.hh.rh > 0 then
          begin
            print(374);
            printint(a.hh.rh);
          end;
      end;
      2: if a.int <> 0 then
        begin
          print(375);
          showbox(a.int);
        end;
    end{:219}
;
  end;
end;

File: cpp/showactivities.cpp
#include "showactivities.h"
#include "printchar.h"
#include "printesc.h"
#include "printscaled.h"
#include "printtotals.h"
#include "printmode.h"
#include "println.h"
#include "print.h"
#include "printnl.h"
#include "printint.h"
#include "showbox.h"
#include "xovern.h"

void showactivities(void)
{
	nest[nestptr] = curlist;
	printnl("");
	println();
	for (int p = nestptr; p >= 0; p--)
	{
		int m = nest[p].modefield;
		memoryword a = nest[p].auxfield;
		printnl("### ");
		printmode(m);
		print(" entered at line ");
		printint(abs(nest[p].mlfield));
		if (m == hmode && nest[p].pgfield != 8585216)
		{
			print(" (language");
			printint(nest[p].pgfield%0x1'00'00);
			print(":hyphenmin");
			printint(nest[p].pgfield/4194304);
			printchar(',');
			printint((nest[p].pgfield/0x1'00'00)%64);
			printchar(')');
		}
		if (nest[p].mlfield < 0)
			print(" (\\output routine)");
		if (p == 0)
		{
			if (page_head != pagetail)
			{
				printnl("### current page:");
				if (outputactive)
					print(" (held over for link output)");
				showbox(link(page_head));
				if (pagecontents > 0)
				{
					printnl("total height ");
					printtotals();
					printnl(" goal height");
					printscaled(pagesofar[0]);
					halfword r = link(page_ins_head);
					while (r != page_ins_head)
					{
						println();
						printesc("insert");
						int t = subtype(r);
						printint(t);
						print(" adds ");
						if (count(t) == 1000)
							t = height(r);
						else
							t = xovern(height(r), 1000)*count(t);
						printscaled(t);
						if (type(r) == vlist_node)
						{
							halfword q = page_head;
							t = 0;
							do
							{
								q = link(q);
								if (type(q) == ins_node && subtype(q) == subtype(r))
									t++;
							} while (q != llink(r));
							print(", #");
							printint(t);
							print(" might split");
						}
						r = link(r);
					}
				}
			}
			;
			if (link(contrib_head))
			printnl("### recent contributions:");
		}
		showbox(link(nest[p].headfield));
		switch (abs(m)/101)
		{
			case 0:
				printnl("prevdepth ");
				if (a.int_ <= -0x1'00'00*1000)
					print("ignored");
				else
					printscaled(a.int_);
				if (nest[p].pgfield)
				{
					print(", prevgraf ");
					printint(nest[p].pgfield);
					print(" line");
					if (nest[p].pgfield != 1)
						printchar('s');
				}
				break;
			case 1:
				printnl("spacefactor ");
				printint(a.hh.lh);
				if (m > 0 && a.hh.rh > 0)
				{
					print(", current language ");
					printint(a.hh.rh);
				}
				break;
			case 2: 
				if (a.int_)
				{
					print("this will be denominator of:");
					showbox(a.int_);
				}
		}
	}
}
